File: typescript/client/src/KeyframesClient.ts
import { EventEmitter } from 'stream';
import { IPlatformNotification, Producer } from './Model';
import { GVPlatform } from './GVPlatform';
import fs from 'fs';

export enum PreviewSizeNum {
  Small = 120,
  Medium = 240,
  Large = 480,
}

export const PreviewSize = {
  [PreviewSizeNum.Small]: 'small',
  [PreviewSizeNum.Medium]: 'medium',
  [PreviewSizeNum.Large]: 'large',
};

export class KeyframesClient extends EventEmitter {
  private gvPlatform: GVPlatform | null;
  private subscriptions: string[] = [];
  private readonly folderPath: string;
  private subscriptionInterval: NodeJS.Timer;

  private readonly SUBSCRIPTION_RENEW_TIME_SECONDS = 60;

  constructor(baseURL: string, apiKey: string, folderPath: string) {
    super();
    this.gvPlatform = new GVPlatform(baseURL, apiKey);
    this.folderPath = folderPath;
  }

  public async login(): Promise<boolean> {
    return this.gvPlatform.login();
  }

  public async getProducerAsync(fabricId: string, producerName: string): Promise<Producer> {

    const name = encodeURIComponent(producerName);
    const url = `/cluster/matrix/api/v1/producer/${fabricId}/${name}`;

    const result = await this.gvPlatform.get(url);
    if (result.status === 200) {
      const messages: Producer = { producer: result.data };
      return messages;
    }
    return null;
  }

  public async startNotificationListener(): Promise<boolean> {
    this.gvPlatform.on('notification', this.onNotification);
    return this.gvPlatform.startNotificationListener();
  }

  public addKeyframesSubscription(nodeId: string, flowId: string): string {
    const subscription = `gv.ampp.keyframe.${nodeId}.${flowId}.${PreviewSize[PreviewSizeNum.Small]}`;
    this.subscriptions.push(subscription);
    return subscription;
  }

  public async startKeyframesSubscriptionAsync() {
    await Promise.all(
      this.subscriptions.map(async (subscription: string) => {
        await this.gvPlatform.subscribeToNotification(subscription);
        console.log('Subscribing to', subscription);
      }),
    );
    await this.renewFrameCacheSubscriptions();
    this.subscriptionInterval = setInterval(
      this.renewFrameCacheSubscriptions.bind(this),
      this.SUBSCRIPTION_RENEW_TIME_SECONDS * 1000,
    );
  }

  public async stopKeyframesSubscriptionAsync() {
    await Promise.all(
      this.subscriptions.map(async (subscription: string) => {
        await this.gvPlatform.unsubscribeNotification(subscription);
        console.log('Unsubscribing:', subscription);
      }),
    );
    clearInterval(this.subscriptionInterval);
  }

  private async renewFrameCacheSubscriptions() {
    try {
      return await Promise.all(
        this.subscriptions.map(async (subscription: string) => {
          console.log('Renewing subscription', subscription);
          const parts: string[] = subscription.split('.');
          const topic = `${parts[0]}.${parts[1]}.${parts[2]}.${parts[3]}`;
          const flowId = parts[4];
          await this.sendFlowSubscriptionRequest(topic, flowId);
        }),
      );
    } catch (err) {
      console.log('err', err);
      throw new Error('renewFrameCacheSubscriptions() error');
    }
  }

  private async sendFlowSubscriptionRequest(topic: string, flowId: string) {
    const content = {
      PreviewSize: PreviewSizeNum.Small,
      FlowId: flowId,
    };
    await this.gvPlatform.publishNotification(topic, content);
  }

  private onNotification = (notification: IPlatformNotification) => {
    console.log('Received keyframes notification');
    if (notification.content === 'image/jpeg' && notification.binaryContent) {
      const filePath = `${this.folderPath}/keyframe.jpg`;
      try {
        fs.writeFileSync(filePath, Buffer.from(notification.binaryContent, 'base64'));
        console.log('The file was saved to', filePath);
        console.log(new Date());
      } catch (err) {
        console.log('Error saving the file:', err);
      }
    }
  };
}

File: csharp/client/ampp.control.common/KeyframesClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Net.Mime;
using System.Threading;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Serilog;
using AmppControl.Model;
using ampp.control.common.Model;

namespace ampp.control.common
{
    public class KeyframesClient
    {
        private GVPlatform gvPlatform;

        private List<string> subscriptions = new List<string>();
        private readonly string folderPath;
        private readonly CancellationTokenSource frameCacheSubscriptionCancellationToken = new CancellationTokenSource();

        private const int SUBSCRIPTION_RENEW_TIME_SECONDS = 60;

        public KeyframesClient(string baseURL, string apiKey, string folderPath)
        {
            gvPlatform = new GVPlatform(baseURL, apiKey);
            this.folderPath = folderPath;
        }

        public async Task<bool> LoginAsync()
        {
            bool connected = await this.gvPlatform.LoginAsync();
            if (connected)
            {
                gvPlatform.OnPushNotifyEvent += OnNotification;
                return await this.gvPlatform.StartNotificationsListenerAsync();
            }
            return connected;
        }


        public async Task<Producer> GetProducerAsync(Guid fabricId, string producerName)
        {
            string requestUri = $"/cluster/matrix/api/v1/producer/{fabricId}/{producerName}";
            HttpResponseMessage response = await this.gvPlatform.Get(requestUri);

            if (response.IsSuccessStatusCode)
            {
                string body = response.Content.ReadAsStringAsync().Result;
                return JsonConvert.DeserializeObject<Producer>(body);
            }
            return null;
        }

        public void AddKeyframesSubscription(string nodeId, string flowId)
        {
            string subscription = $"gv.ampp.keyframe.{nodeId}.{flowId}.{PreviewSize.Small.ToString().ToLower()}";
            subscriptions.Add(subscription);
        }

        public async Task StartKeyframesSubscriptionAsync()
        {
            foreach (string subscription in subscriptions)
            {
                await gvPlatform.SubscribeToNotification(subscription);
                Console.WriteLine($"Subscribing {subscription}");
            }
            await Task.Run(async () => 
            { 
                await RenewFrameCacheSubscriptions(); 
            }, 
            frameCacheSubscriptionCancellationToken.Token);
        }

        private async Task RenewFrameCacheSubscriptions()
        {
            try
            {
                while (!frameCacheSubscriptionCancellationToken.Token.IsCancellationRequested)
                {
                    foreach (string subscription in subscriptions)
                    {
                        Console.WriteLine($"Renewing subscription: {subscription}");
                        string[] parts = subscription.Split('.');
                        string topic = $"{parts[0]}.{parts[1]}.{parts[2]}.{parts[3]}";
                        string flowId = parts[4];
                        await SendFlowSubscriptionRequest(topic, flowId);
                    }
                    await Task.Delay(SUBSCRIPTION_RENEW_TIME_SECONDS * 1000);
                }
            }
            catch (Exception ex)
            {
                Log.Error("Error while subscribing to keyframes!", ex);
                throw;
            }
        }

        public async Task StopKeyframesSubscriptionAsync()
        {
            frameCacheSubscriptionCancellationToken.Cancel();
            foreach (string subscription in subscriptions)
            {
                await gvPlatform.UnsubscribeFromNotification(subscription);
                Console.WriteLine($"Unsubscribing {subscription}");
            }
        }

        // This method tells the framecache to generate keyframes
        private Task SendFlowSubscriptionRequest(string topic, string flowId)
        {
            KeyframesSubscriptionNotification notification = new KeyframesSubscriptionNotification()
            {
                PreviewSize = PreviewSize.Small,
                FlowId = flowId
            };
            return gvPlatform.PushNotificationAsync(topic, JObject.FromObject(notification));
        }

        private void OnNotification(object sender, PushNotification e)
        {
            Console.WriteLine("Received keyframes notification");
            if(e.Content == MediaTypeNames.Image.Jpeg && e.BinaryContent is not null)
            {
                string filePath = $"{folderPath}\\keyframe.jpg";
                File.WriteAllBytes(filePath, e.BinaryContent);
                Console.WriteLine($"Saved keyframe as {filePath}");
                Console.WriteLine(DateTime.Now.ToString());
            }
            else if (e.Content == "Removed")
            {
                Console.WriteLine("Notification is removed");
            }
            else
            {
                throw new Exception("Error handling keyframes notification");
            }
        }
    }
}
